reconcilliation-test/cf_billing_reconcilliation.go

```diff
@@ -0,0 +1,197 @@
+				results.OnlyInSecondSet[k] = true
+			}
+		}
+	}
+	return results
+
+}
+func join_guids(guids map[string]bool) string {
+
+	var keys []string
+
+	for k, _ := range guids {
+		if k != "" {
+			keys = append(keys, k)
+		}
+	}
+	return (strings.Join(keys, ", "))
+}
+
+func main() {
+	//Do the billing api request
+	billingAPIURL, err := url.Parse(os.Getenv("BILLING_API_URL"))
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	billingAPIURL.Path = "/billable_events"
+
+	q := billingAPIURL.Query()
+	q.Set("range_start", time.Now().AddDate(0, 0, -2).Format("2006-01-02"))
+	q.Set("range_stop", time.Now().Format("2006-01-02"))
+	billingAPIURL.RawQuery = q.Encode()
+	billingAPIURL.ForceQuery = true
+
+	req, err := http.NewRequest("GET", billingAPIURL.String(), nil)
+	if err != nil {
+		fmt.Println(err)
+	}
+	headers := req.Header
+	headers.Set("Authorization", fmt.Sprintf("Bearer %s", os.Getenv("CF_BEARER_TOKEN")))
+	req.Header = headers
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	var billableEvents []eventio.BillableEvent
+
+	err = json.Unmarshal([]byte(body), &billableEvents)
+	if err != nil {
+		fmt.Println(err)
+	}
+	billing_service_instance_guids := map[string]bool{}
+	cf_service_instance_guids := map[string]bool{}
+
+	for i := 0; i < len(billableEvents); i++ {
+		// Insert all services found in billing into a map that can be compared later
+		if billableEvents[i].ResourceType == "service" {
+			billing_service_instance_guids[billableEvents[i].ResourceGUID] = true
+		}
+	}
+
+	//Exit out if we have more than 1000 guids as we don't currently paginate the cf query
+	if len(billing_service_instance_guids) > 1000 {
+
+		fmt.Println("There are more than 1000 service instances and the scripts need to be updated to cope with this scenario")
+		os.Exit(1)
+	}
+
+	cfAPIURL, err := url.Parse(os.Getenv("CF_API_URL"))
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	cfAPIURL.Path = "/v3/service_instances"
+
+	q = cfAPIURL.Query()
+	//Set up the query for cf things that have been created for four hours to avoid smoke tests and similar that won't have been picked up by billing yet
+	q.Set("per_page", "1000")
+	q.Set("created_ats[lt]", time.Now().Add(time.Hour * -4).Format("2006-01-02T15:04:05Z"))
+	cfAPIURL.RawQuery = q.Encode()
+	cfAPIURL.ForceQuery = true
+
+	req, err = http.NewRequest("GET", cfAPIURL.String(), nil)
+	req.Header = headers
+	if err != nil {
+		fmt.Println(err)
+	}
+        // Do the request to the cf api
+	cfHTTPClient := &http.Client{}
+	resp, err = cfHTTPClient.Do(req)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	body, err = io.ReadAll(resp.Body)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	var serviceInstances listV3ServiceInstancesResponse
+
+	err = json.Unmarshal([]byte(body), &serviceInstances)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	for i := 0; i < len(serviceInstances.Resources); i++ {
```

Comment:
Why not `for _, serviceInstance := range serviceInstances` here?

Reply:
Fair point I'll update